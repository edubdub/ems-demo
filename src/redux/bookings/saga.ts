import { takeLatest, call, put } from 'redux-saga/effects'
import { actions, actionCreators } from './actions'
import * as bookingService from '../../services/bookings'
export function * bookingSaga() {
  yield takeLatest(actions.LOAD_BOOKINGS, loadBookings)

}

export function * loadBookings () {
  const bookings = yield call(bookingService.requestBookings)
  yield put(actionCreators.setBookings(bookings))
}
