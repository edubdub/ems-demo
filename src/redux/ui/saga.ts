import { call, take, all, fork, select, cancel, put } from 'redux-saga/effects'
import { eventChannel, delay } from 'redux-saga'
import { actions, actionCreators } from './actions'
import selectors from './selectors'
import tween from '../../utils/tween'
import { List } from 'immutable'
const throttle = require('lodash.throttle')
export function * globalEventsSaga() {
  // wait for the nav bar height, and for at least one booking date to be rendered
  yield all[
    yield take(actions.SET_NAV_BAR_HEIGHT),
    yield take(actions.SET_BOOKING_DATE_WINDOW_POSITION)
  ]
  // there could be a bunch of bookings rendering at once, and it's jarring to have it load before
  // they have even appeared from the fade affect
  yield delay(500)
  yield scrollToSelectedDate()
  yield all([
    scrollSaga()
  ])
}

export function * scrollSaga () {
  while (true) {
    const watchScrollTask = yield fork(watchScroll)
    yield take(actions.USER_SET_NAV_DATE)
    cancel(watchScrollTask)
    yield scrollToSelectedDate()
  }
}

export function * scrollToSelectedDate () {
  const positionOfSelectedDate: number = yield select(selectors.positionOfSelectedDate)
  const height = yield select(selectors.navBarHeight)
  yield call(tween, window.scrollY, positionOfSelectedDate - height, 1000, (step: number) => {
    window.scrollTo(0, step)
  })
}

export function * watchScroll () {
  while (true) {
    try {
      const chan = yield call(scrollChannel)
      while (true) {
        const scrollPosition: {x: number, y: number} = yield take(chan)
        yield setDateFromScrollPosition(scrollPosition.y)
      }
    } finally {
      console.log('watchScroll terminated')
    }
  }
}

function * setDateFromScrollPosition (yScrollPosition: number) {
  const height = yield select(selectors.navBarHeight)
  const paddedYPosition = yScrollPosition + height + 50
  const datePositions: List<{bookingDate: Date, position: number}> = yield select(selectors.bookingDatePositions)
  const closestDatePosition = datePositions.reverse().find(datePosition => !!datePosition && paddedYPosition > datePosition.position)
  if (closestDatePosition) {
    yield put(actionCreators.setNavDate(closestDatePosition.bookingDate))
  }
}

const scrollChannel = () => {
  return eventChannel((emitter: (input: any) => any) => {
    const listener = throttle((event: Event) => { emitter({x: window.scrollX, y: window.scrollY}) }, 100, {leading: false, trailing: true})
    window.addEventListener('scroll', listener)
    return () => window.removeEventListener('scroll', listener)
  })
}
