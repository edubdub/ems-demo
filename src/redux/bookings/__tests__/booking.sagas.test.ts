import * as sagas from '../saga'
describe('saga', () => {
  // If I had more time for this, testing something things by more than snapshots would be better
  describe('load bookings', () => {
    const loadBookings = sagas.loadBookings
    it('it should set loading true if true is passed in', () => {
      const iterator = loadBookings(true)
      let result
      while (!result || !result.done) {
        result = iterator.next()
        expect(result).toMatchSnapshot('loadBookings with loading true, step: ')
      }
    })
    it('it should skip loading if false is passed in', () => {
      const iterator = loadBookings(false)
      let result
      while (!result || !result.done) {
        result = iterator.next()
        expect(result).toMatchSnapshot('loadBookings with loading false, step: ')
      }
    })
  })
})
