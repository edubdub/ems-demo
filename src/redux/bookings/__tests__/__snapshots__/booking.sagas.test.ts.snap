// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadBookings with loading false, step:  1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "CALL": Object {
      "args": Array [],
      "context": null,
      "fn": [Function],
    },
  },
}
`;

exports[`loadBookings with loading false, step:  2`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "payload": Immutable.List [
],
        "type": "SET_BOOKINGS",
      },
      "channel": null,
    },
  },
}
`;

exports[`loadBookings with loading false, step:  3`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`loadBookings with loading true, step:  1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "payload": true,
        "type": "SET_BOOKING_LOADING",
      },
      "channel": null,
    },
  },
}
`;

exports[`loadBookings with loading true, step:  2`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "CALL": Object {
      "args": Array [],
      "context": null,
      "fn": [Function],
    },
  },
}
`;

exports[`loadBookings with loading true, step:  3`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "payload": Immutable.List [
],
        "type": "SET_BOOKINGS",
      },
      "channel": null,
    },
  },
}
`;

exports[`loadBookings with loading true, step:  4`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "payload": false,
        "type": "SET_BOOKING_LOADING",
      },
      "channel": null,
    },
  },
}
`;

exports[`loadBookings with loading true, step:  5`] = `
Object {
  "done": true,
  "value": undefined,
}
`;
